plugin/tsig/tsig.go

```diff
@@ -0,0 +1,141 @@
+package tsig
+
+import (
+	"context"
+	"encoding/binary"
+	"encoding/hex"
+	"time"
+
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/plugin/pkg/log"
+	"github.com/coredns/coredns/request"
+
+	"github.com/miekg/dns"
+)
+
+// TSIGServer verifies tsig status and adds tsig to responses
+type TSIGServer struct {
+	Zones   []string
+	secrets map[string]string // [key-name]secret
+	types   qTypes
+	all     bool
+	Next    plugin.Handler
+}
+
+type qTypes map[uint16]struct{}
+
+// Name implements plugin.Handler
+func (t TSIGServer) Name() string { return pluginName }
+
+// ServeDNS implements plugin.Handler
+func (t *TSIGServer) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	var err error
+	state := request.Request{Req: r, W: w}
+	if z := plugin.Zones(t.Zones).Matches(state.Name()); z == "" {
+		return plugin.NextOrFailure(t.Name(), t.Next, ctx, w, r)
+	}
+
+	var tsigRR = r.IsTsig()
+	rcode := dns.RcodeSuccess
+	if !t.tsigRequired(state.QType()) && tsigRR == nil {
+		return plugin.NextOrFailure(t.Name(), t.Next, ctx, w, r)
+	}
+
+	if tsigRR == nil {
+		log.Debugf("rejecting '%s' request without TSIG\n", dns.TypeToString[state.QType()])
+		rcode = dns.RcodeRefused
+	}
+
+	// wrap the response writer so the response will be TSIG signed.
+	w = &restoreTsigWriter{w, r, tsigRR}
+
+	if err = w.TsigStatus(); err != nil {
+		log.Debugf("TSIG validation failed: %v %v", dns.TypeToString[state.QType()], err)
+		rcode = dns.RcodeNotAuth
+		switch err {
+		case dns.ErrSecret:
+			tsigRR.Error = dns.RcodeBadKey
+			// See RFC-2845: 4.7. Special considerations for forwarding servers
```

Comment:
RFC8945 Section 5.5 would be a better reference.  I'd also note that the entire (raw) DNS message would have to be kept and passed to the forwarding plugin; if, for example, the incoming message doesn't compress names, the forwarded message must not compress names either.  I suspect we'd need some more architectural extension to make it possible (like allowing plugins to access raw the incoming DNS message in addition to the "unpacked" `dns.Msg`), rather than just not using this plugin.

Reply:
This was more of a note to self that I forgot to remove...
  
Thanks though, I didn't think of the compression case. I'll add that to the "Bugs" section I'm working on.